From b70097a3f3f7729006b7b00683a8d1b8db2eb95d Mon Sep 17 00:00:00 2001
From: Duy Bao Dang <duy.dang.yb@renesas.com>
Date: Wed, 30 Dec 2020 11:40:22 +0700
Subject: [PATCH 213/440] media: soc_camera: ov5650: Properly release resources
 before exit

This patch improves ov5640_remove to properly release resources before
driver exit. This is to fix the issue of detecting device after board
reboot that may happen randomly.

Signed-off-by: Duy Bao Dang <duy.dang.yb@renesas.com>
Signed-off-by: Hao Bui <hao.bui.yg@renesas.com>
---
 drivers/media/i2c/soc_camera/ov5640.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/media/i2c/soc_camera/ov5640.c b/drivers/media/i2c/soc_camera/ov5640.c
index 11c304d..80dbc2c 100644
--- a/drivers/media/i2c/soc_camera/ov5640.c
+++ b/drivers/media/i2c/soc_camera/ov5640.c
@@ -952,12 +952,16 @@ err_probe:
 static int ov5640_remove(struct i2c_client *client)
 {
 	struct v4l2_subdev *sd = i2c_get_clientdata(client);
+	struct ov5640 *ov5640 = to_ov5640(sd);
 
 	v4l2_dbg(1, debug, sd,
 			"ov5640.c: removing ov5640 adapter on address 0x%x\n",
 			client->addr << 1);
 
 	v4l2_device_unregister_subdev(sd);
+	v4l2_async_unregister_subdev(&ov5640->sd);
+	media_entity_cleanup(&ov5640->sd.entity);
+	v4l2_ctrl_handler_free(&ov5640->hdl);
 	kfree(to_ov5640(sd));
 	return 0;
 }
-- 
2.7.4

